inc/saiipsec.h

```diff
@@ -0,0 +1,1047 @@
+/**
+ * Copyright (c) 2014 Microsoft Open Technologies, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *    not use this file except in compliance with the License. You may obtain
+ *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    saiipsec.h
+ *
+ * @brief   This module defines SAI IPsec interface
+ */
+
+#if !defined (__SAIIPSEC_H_)
+#define __SAIIPSEC_H_
+
+#include <saitypes.h>
+
+/**
+ * @defgroup SAIIPSEC SAI - IPsec specific API definitions
+ *
+ * @{
+ */
+
+/**
+ * @brief IPsec direction types
+ * For PHY ASIC Egress is system to line direction and ingress is the opposite.
+ */
+typedef enum _sai_ipsec_direction_t
+{
+    SAI_IPSEC_DIRECTION_EGRESS,
+    SAI_IPSEC_DIRECTION_INGRESS,
+} sai_ipsec_direction_t;
+
+/**
+ * @brief IPsec cipher suite types
+ */
+typedef enum _sai_ipsec_cipher_t
+{
+    SAI_IPSEC_CIPHER_AES128_GCM16,
+    SAI_IPSEC_CIPHER_AES256_GCM16,
+    SAI_IPSEC_CIPHER_AES128_GMAC,
+    SAI_IPSEC_CIPHER_AES256_GMAC,
+} sai_ipsec_cipher_t;
+
+/**
+ * @brief IPsec SA sequence number status type
+ */
+typedef enum _sai_ipsec_sa_status_t
+{
+    /** SA sequence number above lower of 2 watermarks */
+    SAI_IPSEC_SA_STATUS_SN_ABOVE_LOW_WATERMARK,
```

Comment:
Is this intended to reflect the SA soft and hard lifetime limits from RFC4301? If so then I think there is something missing by including SN in the name. [RFC4301 section 4.4.2.1](https://datatracker.ietf.org/doc/html/rfc4301#section-4.4.2.1) describes lifetime limits being specified in either time or a byte count or both. Using time doesn't really make sense for a switch, so the lifetime support here should really include support for byte count hard and soft limits instead of just packet count limits which is implied by the SN usage.

Reply:
This is why in earlier conversation sought I sought input about the per-SA notification conditions (bot got no response).  The rationale for my choice was the per-SA time bound is normally tracked by switch-software (NOS) and not by the IPsec engine.  With such an approach, the time-limit is not part of the notification from the IPsec engine.

Reply:
I agree that the time based lifetime would be tracked by software and not by the IPsec engine and so out of scope for here. But that still leaves hard and soft limits (per-SA) based on byte counts. I believe these should be in scope and require status enums that don't relate to the SN.

Reply:
Sorry I missed your point - and thanks for pointing it out.  I will replace the SN threshold with byte threshold.